src/Storages/StorageReplicatedMergeTree.cpp

```diff
@@ -1364,6 +1364,19 @@ void StorageReplicatedMergeTree::checkParts(bool skip_sanity_checks)
 
     paranoidCheckForCoveredPartsInZooKeeperOnStart(expected_parts_vec, parts_to_fetch);
 
+    ActiveDataPartSet empty_unexpected_parts_set(format_version);
+    for (const auto & part : parts)
+    {
+        if (part->rows_count || part->getState() != MergeTreeDataPartState::Active || expected_parts.contains(part->name))
```

Comment:
So empty part after unsuccessful TTL merge, or failed-to-remove empty part should pass this condition.

Reply:
> failed-to-remove empty part

It's okay (why would we care about unexpected parts covered by an empty part that we were going to remove?)

> empty part after unsuccessful TTL merge

And it's not. Nice catch. There are two options:
1. Don't set `remove_tmp_policy` to `REMOVE_BLOBS` on server start (will leave some garbage in S3 in rare cases, but I guess it's okay). Otherwise, we can accidentally remove some shared blobs if we fail to commit a fetched empty part (it will not lead to data loss though, but will trigger "no such key" errors). And it should be okay without touching `remove_tmp_policy` because the source parts should be expected if we fail to commit something
2. Find a way to distinguish different kinds of empty parts. Maybe we can use 99999999 as level for such parts. Or don't create empty parts on disk at all, find another way to mark that some parts range was dropped

Reply:
I think I'll choose the first option